### PR TITLE
fix: async issues solved

### PR DESCRIPTION
**Pull request by Quasimurdock**
merge mkanki-dev to mkanki

### CODE REVIEW

File: src/mkanki.js [hunk 1]
```diff
@@ -1,149 +1,106 @@
...
+            } else {
+              try {
+                const $ = cheerio.load(data);
+                let front = "";
+                $(".pos-header.dpos-h")
+                  .toArray()
+                  .forEach((ele) => {
+                    front += $(ele).html() + "<hr color='grey'>";
+                  });
+                const back = $("body").html();
```
Comment (serifold) [inline src/mkanki.js:63]: Some HTML file could be empty, we could throw a warn:

```js
                if (!front || !back) {
                  writeToLog(`[WARN] The file ${file} is empty`);
                }
```

Or just `reject()` the blank file's generation. I prefer reject it, cause generating a blank card seems meaningless.

File: src/mkanki.js [hunk 2]
```diff
@@ -1,149 +1,106 @@
...
-    //           writeToLog(`[ERR] Error processing ${file}: ${err.message}`);
-    //         }
-    //       });
-    //     }
-    //   });
-    // }
-    // addNote()
-  });
+    const result = await Promise.allSettled(files.map(addNote));
+    console.log(`[INFO] addNotes() result: ${JSON.stringify(result)}`);
```
Comment (serifold) [inline src/mkanki.js]: These results are too long, we don't need all of them. We only care about errors.

```js
    result.forEach((e, i) => {
      if (e.status === "rejected") {
        writeToLog(
          `[ERR] ADD ${files[i].match(/^(.+)\.html$/)[1]} FAILED: ${e.reason}`
        );
      }
    });
```

File: src/mkanki.js [hunk 3]
```diff
@@ -1,149 +1,106 @@
...
+                  const tagsDictinct = [...new Set(tags)].sort();
+                  tagResult = tagsDictinct;
+                }
+                deck.addNote(
+                  model.note(
+                    [file.match(/^(.+)\.html$/)[1], front, back],
+                    tagResult
+                  )
+                );
+                console.log(`[INFO] ${++cnt}/${files.length} CURRENT WORD: ${file.match(/^(.+)\.html$/)[1]}`);
```
Comment (serifold) [inline src/mkanki.js]: If we put `common.css` in `src/output/html`, then the number of HTML files should minus 1:

```js
                console.log(
                  `[INFO] ${++cnt}/${files.length - 1} CURRENT WORD: ${
                    file.match(/^(.+)\.html$/)[1]
                  }`
                );
```

Or we could move `common.css` to a `src/css` folder, it isn't a HTML file after all.

File: src/utils/logger.js [hunk 4]
```diff
@@ -1,8 +1,13 @@
...
-  console.log(message);
-  const logMessage = `[${new Date().toLocaleString()}] ${message}\n`;
+  const caller = arguments.callee.caller;
+  let callerName = caller.name;
+  if (caller && callerName === "") {
+    callerName = "anonymous arrow function";
+  }
+  const logMessage = `[${new Date().toLocaleString()}] - ${message} - [${callerName}]\n`;
+  console.log(logMessage);
   fs.appendFile("src/output/log/" + LOG_FILE, logMessage, (err) => {
```
Comment (serifold) [inline src/utils/logger.js]: The `src/output/log/log_html2notes.txt` doesn't exist by default now. We should create a log folder, or also create a blank `log_html2notes.txt` in the log folder.

```js
const LOG_FILE_PATH = "src/output/log/" + LOG_FILE;

  if (!fs.existsSync(LOG_FILE_PATH)) {
    fs.writeFile(LOG_FILE_PATH, "", (err) => {
      if (err) {
        console.log(`[ERR] Failed to create ${LOG_FILE}:`, err);
      }
    });
  }
```

**Review by serifold — APPROVED** [review]

**Comment by Quasimurdock** [comment]
~~@Serifold Hi appreciate ur suggestions.~~

~~Can u give organize and make a PR for them after I merging this one? Thx a lot!~~

Never mind. I'll try to see if I can make a new commit to this PR.

**Comment by serifold** [comment]
> ~@Serifold Hi appreciate ur suggestions.~
> 
> ~Can u give organize and make a PR for them after I merging this one? Thx a lot!~
> 
> Never mind. I'll try to see if I can make a new commit to this PR.

Excuse me, I'm not familiar with github merge, do you mean I could `Merge pull request` first (by default 'Create a merge commit'):

(I thought you could merge too after PR approved, did I figure out something incorrectly?)

then commit these changes I mentioned above to `mkanki-dev`, and make a PR to merge these changes to `mkanki`, finally you'll merge `mkanki` to `main`?

If I understood correctly, sure, I'll do that.

File: src/utils/logger.js (hunk 4 above)
Comment (Quasimurdock) [reply src/utils/logger.js]: > also create a blank log_html2notes.txt in the log folder

`fs.appendFile` actually creates a new file if it doesn't exist:

> Asynchronously append data to a file, creating the file if it does not yet exist.  ([ref](https://nodejs.org/dist/latest-v18.x/docs/api/fs.html#fsappendfilepath-data-options-callback))

But the directory of `log` is needed. I've added in the latest commit.

**Comment by Quasimurdock** [comment]
> I thought you could merge too after PR approved

Yeah. I've pushed a new commit to this PR.

**Pull request merged**